### PR TITLE
Skip invalid grads in outplace foreachs' backward

### LINKED ISSUE #100248: Foreach backward does not handle undefined gradients
For out-of-place foreach functions, the case where not all grad_outputs are defined needs to be handled properly. Currently we would raise an error in this case:

Repro:
```
a = torch.tensor(1., requires_grad=True)
b = torch.tensor(2., requires_grad=True)
out1, out2 = torch._foreach_sigmoid([a, b])
out1.backward()
```
Error:
`RuntimeError: Expected a proper Tensor but got None (or an undefined Tensor in C++) for argument #0 'grad_output`

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @Lezcano @Varal7 @crcrpar @mcarilli @ngimel

### PR DESCRIPTION
**Pull request by crcrpar**
Fixes #100248

### CODE REVIEW

File: tools/autograd/gen_autograd_functions.py [hunk 1]
```diff
@@ -109,7 +109,12 @@
   std::vector<Tensor> grad_result;
   grad_result.reserve(grads.size());
   for (const auto & i : c10::irange(grads.size())) {
-    grad_result.emplace_back(${derivative});
+    const auto& grad = grads[i];
+    if (grad.defined()) {
+      grad_result.emplace_back(${derivative});
+    } else {
+      grad_result.emplace_back(Tensor());
```
Comment (crcrpar) [inline tools/autograd/gen_autograd_functions.py:116]: Is this okay when creating a Tensor whose `.defined()` returns `false`?

File: test/test_foreach.py [hunk 2]
```diff
@@ -878,6 +878,16 @@ def test_inplace_foreach_leaf_check_and_grad_fn(self, device, dtype, op):
         self.assertIsNotNone(tensors[0].grad_fn)
         self.assertIsNone(tensors[1].grad_fn)
 
+    @onlyCUDA
+    @ops(foreach_unary_op_db, dtypes=(torch.float,))
```
Comment (crcrpar) [inline test/test_foreach.py:882]: I'll cover the other foreach op db's if the direction looks okay

File: tools/autograd/gen_autograd_functions.py (hunk 1 above)
Comment (soulitzer) [reply tools/autograd/gen_autograd_functions.py:115]: Yeah its okay to return undefined tensors.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 3]
```diff
@@ -88,7 +88,11 @@ void copy_range(variable_list& out, IndexRange range, at::ArrayRef<Tensor> t) {
   TORCH_CHECK(
       range.second - range.first == t.size(),
       "inconsistent range for TensorList output");
-  std::copy(t.begin(), t.end(), out.begin() + range.first);
+  std::copy_if(
```
Comment (soulitzer) [inline torch/csrc/autograd/FunctionsManual.cpp:91]: This is what you were referring to above? I think it is okay without this change.

**Review by soulitzer — APPROVED** [review]
Thanks!

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 3 above)
Comment (crcrpar) [reply torch/csrc/autograd/FunctionsManual.cpp]: yeah, you're right. `copy_if` was my first too optimistic approach before just creating undefined Tensors.

**Review by albanD — APPROVED** [review]
Thanks for the quick fix!

**Comment by crcrpar** [comment]
could anyone use your strong hammer to merge?
I do think the failure has nothing to do with this change

**Comment by crcrpar** [comment]
> ## Merge failed
> **Reason**: Not merging any PRs at the moment because there is a merge blocking https://github.com/pytorch/pytorch/labels/ci:%20sev issue open at: #100325
> 
> Details for Dev Infra team

I don't quite understand what was the problem

**Comment by albanD** [comment]
> I don't quite understand what was the problem

@ZainRizvi is there doc on the SEV process for CI?

the short version is that when CI is in too bad a state, we can use such a system to block all merges until we go back into a stable state.

**Comment by ZainRizvi** [comment]
The Sev that blocked your PR in particular was also not a real sev. Last week someone discovered that anyone could create an issue with the right shape and have it be considered as sev and created a fake one.  We've since patched our system to only consider sevs filed by trusted contributors to be "real" sevs

See this PR for more details
https://github.com/pytorch/test-infra/pull/4092

**Comment by albanD** [comment]
I was curious where do we have documentation explaining what is the "sev process"?

**Comment by ZainRizvi** [comment]
I believe that's only documented internally (we try to make the external sev issues self-explanatory). I can send you the link

**Comment by albanD** [comment]
Zain added it to the wiki! https://github.com/pytorch/pytorch/wiki/What-is-considered-a-SEV%3F 
Hopefully that clarifies things!

**Pull request closed without merging**